### ISSUE TITLE
main,py,code

### ISSUE THREAD

**Issue by vijayr2004** [vijayr2004/voic-based-aythentication#9]
import os
import numpy as np
from flask import Flask, request, jsonify
from pydub import AudioSegment
import librosa
from sklearn.svm import SVC
import joblib

app = Flask(__name__)
UPLOAD_FOLDER = 'uploads'
os.makedirs(UPLOAD_FOLDER, exist_ok=True)

# Convert mp3 to wav
def convert_mp3_to_wav(mp3_path, wav_path):
    audio = AudioSegment.from_mp3(mp3_path)
    audio.export(wav_path, format="wav")

# Extract MFCC features
def extract_features(file_path):
    y, sr = librosa.load(file_path, sr=None)
    mfccs = librosa.feature.mfcc(y=y, sr=sr, n_mfcc=13)
    return np.mean(mfccs.T, axis=0)

# Load dataset
def load_dataset(data_folder):
    X = []
    y = []
    for speaker in os.listdir(data_folder):
        speaker_path = os.path.join(data_folder, speaker)
        if not os.path.isdir(speaker_path):
            continue
        for file in os.listdir(speaker_path):
            if file.endswith(".mp3"):
                mp3_path = os.path.join(speaker_path, file)
                wav_path = mp3_path.replace(".mp3", ".wav")
                convert_mp3_to_wav(mp3_path, wav_path)
                features = extract_features(wav_path)
                X.append(features)
                y.append(speaker)
                os.remove(wav_path)
    return np.array(X), np.array(y)

# Train model (once)
def train_model():
    X, y = load_dataset("voice_data")
    model = SVC(kernel='linear', probability=True)
    model.fit(X, y)
    joblib.dump(model, "voice_model.pkl")
    print("✅ Model trained and saved.")
    return model

# Load or train model
if os.path.exists("voice_model.pkl"):
    model = joblib.load("voice_model.pkl")
else:
    model = train_model()

# Authentication route
@app.route("/authenticate", methods=["POST"])
def authenticate():
    file = request.files["file"]
    if not file:
        return jsonify({"error": "No file uploaded"}), 400

    mp3_path = os.path.join(UPLOAD_FOLDER, file.filename)
    wav_path = mp3_path.replace(".mp3", ".wav")
    file.save(mp3_path)

    try:
        convert_mp3_to_wav(mp3_path, wav_path)
        features = extract_features(wav_path)
        user = model.predict([features])[0]
        confidence = model.predict_proba([features]).max()

        os.remove(mp3_path)
        os.remove(wav_path)

        return jsonify({"user": user, "confidence": round(confidence, 2)})
    except Exception as e:
        return jsonify({"error": str(e)}), 500

if __name__ == "__main__":
    app.run(debug=True)
